File: server/hwapi/data_models/data_validators/devices.py
# Copyright 2024 Canonical Ltd.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License version
# 3, as published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Written by:
#        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>


from pydantic import BaseModel

from hwapi.data_models.enums import DeviceStatus


class AudioValidator(BaseModel):
    """Validator model for audio devices"""

    model: str
    vendor: str
    identifier: str


class BiosValidator(BaseModel):
    vendor: str
    version: str
    revision: str | None
    firmware_revision: str | None
    release_date: str | None


class BoardValidator(BaseModel):
    manufacturer: str
    product_name: str
    version: str


class ChassisValidator(BaseModel):
    chassis_type: str
    manufacturer: str
    sku: str
    version: str


class GPUValidator(BaseModel):
    status: DeviceStatus | None = None
    codename: str | None = None
    manufacturer: str
    version: str
    identifier: str


class NetworkAdapterValidator(BaseModel):
    """Validator for ethernet network adapters"""

    bus: str
    identifier: str
    model: str
    vendor: str
    capacity: int


class PCIPeripheralValidator(BaseModel):
    status: DeviceStatus | None = None
    pci_id: str
    name: str
    vendor: str


class ProcessorValidator(BaseModel):
    identifier: list[int] | None
    frequency: int  # in MHz
    manufacturer: str
    version: str


class USBPeripheralValidator(BaseModel):
    status: DeviceStatus | None = None
    usb_id: str
    name: str
    vendor: str


class VideoCaptureValidator(BaseModel):
    """Validator model for video capture devices"""

    status: DeviceStatus | None = None
    model: str
    vendor: str
    identifier: str


class WirelessAdapterValidator(BaseModel):
    """Validator model for wireless network adapters"""

    status: DeviceStatus | None = None
    model: str
    vendor: str
    identifier: str

File: client/hwlib/src/models/devices.rs
/* Copyright 2024 Canonical Ltd.
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 3, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Written by:
 *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
 */

use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct Audio {
    pub identifier: String,
    pub model: String,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct Bios {
    pub firmware_revision: Option<String>,
    pub release_date: Option<String>,
    pub revision: Option<String>,
    pub vendor: String,
    pub version: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Default)]
pub struct Board {
    pub manufacturer: String,
    pub product_name: String,
    pub version: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct Chassis {
    pub chassis_type: String,
    pub manufacturer: String,
    pub sku: String,
    pub version: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct GPU {
    pub codename: Option<String>,
    pub identifier: String,
    pub manufacturer: String,
    pub status: Option<DeviceStatus>,
    pub version: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct NetworkAdapter {
    pub bus: String,
    pub capacity: i32,
    pub identifier: String,
    pub model: String,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct PCIPeripheral {
    pub pci_id: String,
    pub name: String,
    pub status: Option<DeviceStatus>,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct Processor {
    pub identifier: Option<[u8; 8]>,
    pub frequency: u64,
    pub manufacturer: String,
    pub version: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct USBPeripheral {
    pub usb_id: String,
    pub name: String,
    pub status: Option<DeviceStatus>,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct VideoCapture {
    pub identifier: String,
    pub model: String,
    pub status: Option<DeviceStatus>,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub struct WirelessAdapter {
    pub identifier: String,
    pub model: String,
    pub status: Option<DeviceStatus>,
    pub vendor: String,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
pub enum DeviceStatus {
    Working,
    Breaking,
}
